File: trie.cc
#include <cstdio>
#include <cstdlib>
#include <cstring>
#include <string>
#include <vector>

using namespace std;

const int ALPHABET_SIZE = 26;

class Trie {
    private:
        struct TrieNode {
            int value;
            TrieNode *children[ALPHABET_SIZE];
            TrieNode() { 
                value = 0;
                for(int i = 0; i < ALPHABET_SIZE; i++) {
                    children[i] = NULL;
                }
            }
        };
        TrieNode *root;
        int count;

    public:
        Trie() {
            root = new TrieNode;
            count = 0;
        }

        void insert(string key) {
            int level;
            int length = key.length();
            int index;
            TrieNode *pCrawl;

            count++;
            pCrawl = root;
            for( level = 0; level < length; level++ ) {
                index = key[level] - 'a';
                if( pCrawl->children[index] ) 
                    pCrawl = pCrawl->children[index];
                else {
                    pCrawl->children[index] = new TrieNode;
                    pCrawl = pCrawl->children[index];
                }
            }
            pCrawl->value = count;
        }

        bool search(string key) {
            int length = key.length();
            int index;

            TrieNode *pCrawl;
            pCrawl = root;
            for(int level = 0; level < length; level++ ) {
                index = key[level] - 'a';
                if( !pCrawl->children[index] ) return 0;
                pCrawl = pCrawl->children[index];
            }
            return (pCrawl!=NULL && pCrawl->value);
        }

        bool leafNode(TrieNode *pNode) {
            return (pNode->value != 0);
        }

        bool isItFreeNode(TrieNode *pNode) {
            for(int i = 0; i < ALPHABET_SIZE; i++) 
                if( pNode->children[i] ) return 0;
            return 1;
        }

        bool deleteHelper(TrieNode *pNode, string key, int level, int len) {
            if( pNode ) {
                if( level == len ) {
                    if( pNode->value ) {
                        pNode->value = 0;
                        if( isItFreeNode(pNode) ) return true;
                        return false;
                    }
                }
                else {
                    int index = key[level] - 'a';

                    if( deleteHelper(pNode->children[index], key, level+1, len) ) {
                        delete pNode->children[index];
                        return ( !leafNode(pNode) && isItFreeNode(pNode) );
                    }
                }
            }
            return false;
        }

        void deleteKey(string key) {
            int len = key.length(); 
            if( len > 0 ) 
                deleteHelper(root, key, 0, len);
        }
};

int main() {
    vector<string> keys  = {"she", "sells", "sea", "shore", "the", "by", "sheer"};
    Trie tree;

    for(int i = 0; i < keys.size(); i++) {
        tree.insert(keys[i]);
    }
    printf("%s %s\n", "she", tree.search("she") ? "Present in trie" : "Not present in trie");
    tree.deleteKey(keys[0]);
    printf("%s %s\n", "she", tree.search("she") ? "Present in trie" : "Not present in trie");
 
    return 0;
}

File: Kotlin/Trie.kt
class Trie {
    internal class TrieNode(var value: Int = 0) {
        val children: Array<TrieNode?> = Array<TrieNode?>(26) { null }
    }

    private var root: TrieNode = TrieNode()
    private var count = 0

    fun insert(key: String) {
        count ++
        var node = root
        for (ch in key) {
            if (node.children[ch - 'a'] == null) {
                node.children[ch - 'a'] = TrieNode()
            }
            node = node.children[ch - 'a']!!
        }
        node.value = count
    }

    fun search(key: String): Boolean {
        var node: TrieNode? = root
        for (ch in key) {
            if (node == null) return false
            if (node.children[ch - 'a'] == null) return false
            node = node.children[ch - 'a']
        }
        return node != null && node.value > 0
    }

    private fun isFreeNode(node: TrieNode): Boolean {
        for (i in 0 .. 25) {
            if (node.children[i] != null) return false
        }
        return true
    }

    private fun canDelete(node: TrieNode?, key: String, index: Int): Boolean {
        if (node != null) {
            if (index == key.length) {
                if (node.value > 0) {
                    node.value = 0
                    return isFreeNode(node)
                }
            } else {
                if (canDelete(node.children[key[index] - 'a'], key, index + 1)) {
                    node.children[key[index] - 'a'] = null
                    return (node.value == 0 && isFreeNode(node))
                }
            }
        }
        return false
    }

    fun delete(key: String) {
        canDelete(root, key, 0)
    }
}